Make better use of our render_state struct

BastiaanOlij opened the pull request:
This PR right now takes the first step in a restructure to make it easier to maintain some of the internal rendering code especially in light of the soon to add stereoscopic capabilities of the renderer.

Right now a lot of data is copied all through the renderer through dozens of parameters, it makes for hard to maintain code. 
Interestingly enough, the `RenderState` struct already has most of this data but was only used by a handful of methods. It makes far more sense to pass a pointer to this struct around then all the individual members held within. The problem here is that RenderState also contains data we want to keep internal to `RendererSceneRenderRD`.

This PR splits out the data that is normally passed around as parameters to our Forward Clustered and Forward Mobile renders into a separate struct called `RenderDataRD` and removed from `RenderState` where applicable. An instance of this struct is now created and populated in `render_scene` and similar methods and a pointer to this instance is passed around to various methods (not all related methods have been updated yet). 

I've also added a few new members to this struct that were calculated many times throughout the code while simply caching them in the struct suffices (such as the near and far plane values).

BastiaanOlij committed:
Split RenderDataRD struct from RenderState struct to simplify passing our render data around the renderer.

BastiaanOlij commented:
Made some more improvements and also implemented `_setup_environment` to use `render_state`

BastiaanOlij commented:
Renamed `RenderState` to `RenderStateRD` and placed it outside of the `RendererSceneRenderRD` so it can be passed nicely to the GI code (and possibly other places where needed)

BastiaanOlij commented:
Changed a bunch more methods. @akien-mga questioned whether this would have an effect on how well this data is cached. Not sure, I imaging it would stay in cache and preventing a bunch of copying of data can only improve things especially once we add more fields for stereo support?

JFonS reviewed (approved):
Looks like a step in the right direction.

BastiaanOlij commented:
Added it as an extra commit for now (will squash if needed) but I've added a pointer to our environment structure to `RenderStateRD`. Seeing where we parse this struct around we know we're dealing with `RendererSceneEnvironmentRD` and parsing the RID over and over again.

The RID is still part of the structure because we have yet to rewrite some of the code that doesn't use `render_state` yet but the idea is that we'll only have the pointer in `RenderStateRD` at the end.

I think I'm also checking correctly everywhere whether the pointer isn't `NULL` so we shouldn't have anything falling over.

My gut feeling says we should do similar things for some of the other RIDs in this struct but I suggest turning those into separate PRs at some point and focusing purely on finishing the render state code. I just added this in to show my line of thought.

clayjohn commented on servers/rendering/renderer_rd/renderer_scene_environment_rd.h:
```diff
@@ -34,26 +34,50 @@
 	RS::EnvironmentReflectionSource reflection_source = RS::ENV_REFLECTION_SOURCE_BG;
 	Color ao_color;
 
+	Basis get_sky_orientation() const { return sky_orientation; };
```
I dont understand the benefit of exposing getters and setters like this for environment when the properties are also publicly accessible. Right now it looks like a change with no real benefit

BastiaanOlij replied:
Well the idea is to make them private but I'm really just following convention here. I'm all for just accessing the members directly and removing the getters/setters.

BastiaanOlij replied:
Changed this, indeed no need to over complicate this.

clayjohn commented:
Looks fine to me, I have no preference between this and the current code.

BastiaanOlij commented:
Feedback from @reduz on chat:
- Split RenderStateRD into an internal state that holds data we do not want to expose (cluster buffer, shadow data, etc) and a shared state (camera data, lod, etc. Basically all that is passed to `_render_scene`)
- Remove the work around including the environment buffer pointer. This closes the door to creating render implementations with GDNative

akien-mga commented:
> Remove the work around including the environment buffer pointer. This closes the door to creating render implementations with GDNative

To clarify based on the discussion: it's the workaround which was closing the door to creating render implementation with GDNative, not its removal.

BastiaanOlij commented:
> 
> 
> > Remove the work around including the environment buffer pointer. This closes the door to creating render implementations with GDNative
> 
> To clarify based on the discussion: it's the workaround which was closing the door to creating render implementation with GDNative, not its removal.

Hahaha,, The work done around, not workaround... :) Mastery of english sometimes fails me :)

BastiaanOlij commented:
ok should be better now, @reduz let me know what things you think should still be in `RenderState`, most is still in `RenderDataRD` as they are accessed all over the place.

BastiaanOlij commented:
I'm sure there will be more improvements possible but I think I got all the core stuff working properly. Ready for Review I think

reduz commented on servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:
```diff
@@ -1712,13 +1740,18 @@ void RenderForwardClustered::_render_sdfgi(RID p_render_buffers, const Vector3i
 
 	PassMode pass_mode = PASS_MODE_SDF;
-	_fill_render_list(RENDER_LIST_SECONDARY, p_instances, pass_mode, CameraMatrix(), Transform());
+	_fill_render_list(RENDER_LIST_SECONDARY, &render_data, pass_mode);
```
For sub-functions I would leave as-is, I think code is cleaner this way.

akien-mga commented:
Thanks!
